components/ai_chat/ai_chat_api.cc

```diff
@@ -113,40 +149,78 @@ void AIChatAPI::QueryPrompt(ResponseCallback callback,
   dict.Set("top_p", 0.999);
   dict.Set("model", model_name);
   dict.Set("stop_sequences", std::move(stop_sequences));
-  dict.Set("stream", false);
+  dict.Set("stream", true);
 
-  base::flat_map<std::string, std::string> headers;
-  headers.emplace("x-brave-key", BUILDFLAG(BRAVE_SERVICES_KEY));
+  return dict;
+}
+
+void AIChatAPI::OnDataReceived(base::StringPiece string_piece,
+                               base::OnceClosure resume) {
+  std::vector<std::string> stream_data = base::SplitString(
+      string_piece, "\r\n", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  static constexpr char kDataPrefix[] = "data: {";
 
-  DVLOG(1) << __func__ << " Prompt: " << prompt << "\n";
-  DVLOG(1) << __func__ << " Using model: " << model_name;
+  // Binding a remote will guarantee the order of the messages being sent, so we
+  // must ensure that it is connected.
+  DCHECK(data_decoder_);
+  DCHECK(json_parser_.is_connected());
 
-  api_request_helper_.Request("POST", api_url, CreateJSONRequestBody(dict),
-                              "application/json", std::move(internal_callback),
-                              headers);
+  for (const auto& data : stream_data) {
+    if (!base::StartsWith(data, kDataPrefix)) {
+      continue;
+    }
+
+    std::string json = data.substr(strlen(kDataPrefix) - 1);
 
-  DVLOG(1) << __func__ << " API Request sent\n";
+    data_decoder_->ParseJson(json,
+                             base::BindOnce(&AIChatAPI::OnParseJsonIsolated,
+                                            weak_ptr_factory_.GetWeakPtr()));
+  }
+
+  std::move(resume).Run();
 }
 
-void AIChatAPI::OnGetResponse(ResponseCallback callback,
-                              api_request_helper::APIRequestResult result) {
-  const bool success = result.response_code() == 200;
+void AIChatAPI::OnComplete(bool success) {
+  DCHECK(completion_callback_);
+  DCHECK(current_request_->get());
+
+  int response_code = -1;
 
-  if (!success) {
-    DVLOG(1) << __func__ << " Response from API was not HTTP 200 (Received "
-             << result.response_code() << ")";
+  if (current_request_->get()->ResponseInfo() &&
+      current_request_->get()->ResponseInfo()->headers) {
+    response_code =
+        current_request_->get()->ResponseInfo()->headers->response_code();
   }
 
```

Comment:
nit: I think we can make this simpler.


```suggestion

  if (auto* response_info = current_request_->ResponseInfo();
      response_info && response_info->headers) {
    response_code = response_info->headers->response_code();
  }

```

Reply:
Done.